### PR TITLE
copr: add vectorize add_date_and_string

### LINKED ISSUE #9016: Coprocessor functions migration from non-vec framework
## Feature Request

Coprocessor historically has two implementations. One is based on the vectorized framework, and the other one is based on the plain (non-vec) framework, which was removed recently. A few of the functions in the legacy non-vec framework have not been ported into the vectorized framework yet.

The following is the list of functions that existed in non-vec framework but not in vectorized framework. You may also want to look into the [non-vec framework before it was removed](https://github.com/tikv/tikv/tree/1a88f12ebf50064b992fc5efc7e7f55795210521/components/tidb_query_normal_expr/src).

- [x] AddDateAndDuration
- [ ] AddDateAndString
- [x] AddDatetimeAndDuration
- [x] AddDatetimeAndString
- [x] AddDurationAndDuration
- [x] AddDurationAndString
- [ ] AddTimeDateTimeNull
- [ ] AddTimeDurationNull
- [ ] AddTimeStringNull
- [ ] Compress
- [x] Date
- [ ] DateDiff
- [ ] Instr
- [x] Locate2ArgsUtf8
- [x] Locate3ArgsUtf8
- [ ] Lower
- [x] NullTimeDiff
- [x] Quote
- [x] RegexpSig
- [x] RegexpUtf8Sig
- [ ] RpadUtf8
- [x] SubDateAndDuration
- [ ] SubDateAndString
- [x] SubDatetimeAndDuration
- [ ] SubDatetimeAndString
- [x] SubDurationAndDuration
- [x] SubDurationAndString
- [ ] SubTimeDateTimeNull
- [ ] SubTimeDurationNull
- [ ] Substring2Args
- [ ] Substring2ArgsUtf8
- [ ] Substring3Args
- [ ] Substring3ArgsUtf8
- [ ] Trim2Args
- [ ] TruncateDecimal
- [ ] TruncateUint
- [ ] Uncompress
- [x] WeekWithoutMode
- [x] YearWeekWithMode
- [x] YearWeekWithoutMode

### PR DESCRIPTION
**Pull request by iliveforfun**
Signed-off-by: zhuhaozhuhao notmyfault1996@gmail.com

### What problem does this PR solve?

Issue Number: #9016

Problem Summary:

### What is changed and how it works?

what is changed: add vectorize add_date_and_string and related test.

### Check List 

Tests 

- Unit test

Side effects

### Release note 
- Vectorize add_date_and_string in tikv coprocessor functions

### CODE REVIEW

**Comment by iliveforfun** [comment]
@Xuanwo @zz-jason @zimulala

File: components/tidb_query_expr/src/impl_time.rs [hunk 1]
```diff
@@ -1606,6 +1636,54 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_add_date_and_string() {
+        let mut ctx = EvalContext::default();
+        let cases = vec![
+            (
+                Some("2021-03-26"),
```
Comment (Xuanwo) [inline components/tidb_query_expr/src/impl_time.rs:1644]: Please format code with `cargo fmt`

File: components/tidb_query_expr/src/impl_time.rs [hunk 2]
```diff
@@ -343,7 +343,37 @@ pub fn add_datetime_and_string(
...
+        None => {
+            return ctx
+                .handle_invalid_time_error(Error::overflow(
+                    "DATE",
+                    format!("({} + {})", arg0, arg1),
+                ))
+                .map(|_| Ok(None))?;
+        }
+    };
+    res.set_time_type(TimeType::Date).ok().unwrap();
```
Comment (Xuanwo) [inline components/tidb_query_expr/src/impl_time.rs:371]:
```suggestion
    res.set_time_type(TimeType::Date)?;
```

File: components/tidb_query_expr/src/lib.rs [hunk 3]
```diff
@@ -497,7 +497,6 @@ fn map_expr_node_to_rpn_func(expr: &Expr) -> Result<RpnFnMeta> {
         ScalarFuncSig::JsonRemoveSig => json_remove_fn_meta(),
         ScalarFuncSig::JsonKeysSig => json_keys_fn_meta(),
         ScalarFuncSig::JsonKeys2ArgsSig => json_keys_fn_meta(),
-        ScalarFuncSig::JsonQuoteSig => json_quote_fn_meta(),
```
Comment (Xuanwo) [inline components/tidb_query_expr/src/lib.rs:500]: Why introduce this change?

Comment (iliveforfun) [reply components/tidb_query_expr/src/lib.rs:500]: misoperation...

File: components/tidb_query_expr/src/impl_time.rs (hunk 1 above)
Comment (iliveforfun) [reply components/tidb_query_expr/src/impl_time.rs]: done, ptal

**Comment by Xuanwo** [comment]
@iliveforfun Please follow the https://github.com/pingcap/community/blob/master/contributors/release-note-checker.md to fix the release note job.

File: components/tidb_query_expr/src/impl_time.rs [hunk 4]
```diff
@@ -1606,6 +1634,38 @@ mod tests {
...
     }
 
+    #[test]
+    fn test_add_date_and_string() {
+        let mut ctx = EvalContext::default();
+        let cases = vec![
+            (Some("2021-03-26"), Some("00:00:00"), Some("2021-03-26")),
+            (Some("2021-03-26"), Some("23:59:59"), Some("2021-03-26")),
+            (Some("2021-03-26"), Some("24:00:01"), Some("2021-03-27")),
+            (Some("2021-03-26"), Some("48:00:01"), Some("2021-03-28")),
```
Comment (Xuanwo) [inline components/tidb_query_expr/src/impl_time.rs:1644]: Please add case like:

```
                Some("2018-12-31 23:00:00"),
                Some("1 01:30:30"),
```

and

```
                Some("2019-01-01 01:00:00"),
                Some("-01:01:00"),
```

File: components/tidb_query_expr/src/lib.rs [hunk 5]
```diff
@@ -701,6 +701,7 @@ fn map_expr_node_to_rpn_func(expr: &Expr) -> Result<RpnFnMeta> {
         ScalarFuncSig::SubDurationAndDuration => sub_duration_and_duration_fn_meta(),
         ScalarFuncSig::MakeTime => make_time_fn_meta(),
         ScalarFuncSig::DurationDurationTimeDiff => duration_duration_time_diff_fn_meta(),
+        ScalarFuncSig::AddDateAndString => add_date_and_string_fn_meta(),
```
Comment (Xuanwo) [inline components/tidb_query_expr/src/lib.rs:704]: Maybe we can place this function next to date related functions?

**Comment by sticnarf** [comment]
Hi @iliveforfun, could you please change a bit of the PR description to something like "fix part of #9016"?
The "fix #xxx" syntax will closes that issue, which is not expected.

File: components/tidb_query_expr/src/impl_time.rs (hunk 4 above)
Comment (sticnarf) [reply components/tidb_query_expr/src/impl_time.rs:1644]: Maybe we can assume the first param does not contain time?
Anyway, there should be a case whose first param is a valid date and the second param is an invalid time.

**Review by skyzh — APPROVED** [review]

**Comment by iliveforfun** [comment]
> Hi @iliveforfun, could you please change a bit of the PR description to something like "fix part of #9016"?
> The "fix #xxx" syntax will closes that issue, which is not expected.
done

File: components/tidb_query_expr/src/impl_time.rs (hunk 4 above)
Comment (iliveforfun) [reply components/tidb_query_expr/src/impl_time.rs:1644]: done ptal

File: components/tidb_query_expr/src/lib.rs (hunk 5 above)
Comment (iliveforfun) [reply components/tidb_query_expr/src/lib.rs:704]: done

**Comment by Xuanwo** [comment]
Please don't force push your commits, as this will make it hard to follow the reviews.

**Comment by iliveforfun** [comment]
> Please don't force push your commits, as this will make it hard to follow the reviews.

get it..

**Comment by Xuanwo** [comment]
Please make sure your commits passed all CI checks.

For `idc-jenkins-ci-tikv/check_release_note`, you need to follow https://github.com/pingcap/community/blob/master/contributors/release-note-checker.md , and update PRs description.

For `idc-jenkins-ci/test`, I think you need to do `cargo fmt`.

**Comment by iliveforfun** [comment]
> Please make sure your commits passed all CI checks.
> 
> For `idc-jenkins-ci-tikv/check_release_note`, you need to follow https://github.com/pingcap/community/blob/master/contributors/release-note-checker.md , and update PRs description.
> 
> For `idc-jenkins-ci/test`, I think you need to do `cargo fmt`.

done

**Review by Xuanwo — APPROVED** [review]
